sdk/src/types/block/mana/parameters.rs

```diff
@@ -196,6 +196,59 @@ impl ProtocolParameters {
                 - (c >> mana_parameters.decay_factors_exponent())
         })
     }
+
+    pub fn slots_until_generated(
+        &self,
+        current_slot: impl Into<SlotIndex>,
+        generation_amount: u64,
+        stored_mana: u64,
+        required_mana: u64,
+    ) -> Result<u32, ManaError> {
+        if required_mana == 0 {
+            return Ok(0);
+        }
+        if required_mana > self.mana_parameters().max_mana() {
+            return Err(ManaError::AboveMax {
+                value: required_mana,
+                max: self.mana_parameters().max_mana(),
+            });
+        }
+        let current_slot = current_slot.into();
+        let mut num_slots = 0;
+        let mana_generated_per_epoch = self
+            .mana_parameters()
+            .generate_mana(generation_amount, self.slots_per_epoch());
+        let mut required_mana_remaining = required_mana;
+        loop {
+            // Get the minimum number of slots required to achieve the needed mana (i.e. not including decay)
+            let additional_slots = u32::try_from(
+                (required_mana_remaining as u128 * self.slots_per_epoch() as u128)
+                    / mana_generated_per_epoch.max(1) as u128,
+            )
+            .map_err(|_| ManaError::InsufficientGenerationAmount)?;
+
+            num_slots += additional_slots.max(1);
+            // Get the actual values after than many slots
+            let decayed_mana =
+                stored_mana - self.mana_with_decay(stored_mana, current_slot, current_slot + num_slots)?;
+            let generated_mana =
+                self.generate_mana_with_decay(generation_amount, current_slot, current_slot + num_slots)?;
```

Comment:
I think after this, current_slot should be increased by num_slots, so if running another round in the loop it starts to generate from where it ended and not at the same index again

Reply:
no, the mana is always generated from the original slot to ensure the final check is always correct

Reply:
But that's wrong? Which final check?

Reply:
Ah got it, `num_slots` is updated in the loop, all good